### PR TITLE
[SPARK-12756][SQL] use hash expression in Exchange

### PR DESCRIPTION
**Pull request by cloud-fan**
This PR makes bucketing and exchange share one common hash algorithm, so that we can guarantee the data distribution is same between shuffle and bucketed data source, which enables us to only shuffle one side when join a bucketed table and a normal one.

This PR also fixes the tests that are broken by the new hash behaviour in shuffle.

### CODE REVIEW

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
cc @davies for the pyspark changes

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java [hunk 1]
```diff
@@ -302,16 +300,23 @@ public void testSetOperation() {
 
     Dataset<String> unioned = ds.union(ds2);
     Assert.assertEquals(
-      Arrays.asList("abc", "abc", "foo", "foo", "xyz", "xyz"),
-      sort(unioned.collectAsList().toArray(new String[0])));
+      asSet("abc", "abc", "foo", "foo", "xyz", "xyz"),
```
Comment (nongli) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java]: Why did you change this? The previous version is ordering tolerant no? This changes the test a bit wrt duplicates.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala [hunk 2]
```diff
@@ -324,6 +324,8 @@ private[sql] class DynamicPartitionWriterContainer(
 
   private def bucketIdExpression: Option[Expression] = for {
     BucketSpec(numBuckets, _, _) <- bucketSpec
+  // Use `Pmod` to match the behaviour of `Utils.nonNegativeMod`, which is the function used to
```
Comment (nongli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: I don't think this is a good place for this comment. It needs to be with Pmod and Utils.nonNegativeMod indicating they can't be changed. (Can we also implement Pmod as just calling Utils.nonNegativeMod?) That will make sure they stay in sync.

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java (hunk 1 above)
Comment (cloud-fan) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java]: oh, it's a mistake, thanks for catching it!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: `Pmod` works for all numerics but `Utils.nonNegativeMod` only works for int.
How about we also use `Pmod` in shuffle so that we don't need to depend on `Utils.nonNegativeMod`?

Comment (nongli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: nonNegativeMod seems to be used in core places so making it depend on a SQL class seems wrong. I was just suggesting that the int version of pmod just call utils.nonNegativeMod and you add the comment saying changing this is bad.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: oh, I don't mean to change `Utils.nonNegativeMod` to use `Pmod`, I mean we can use `Pmod(Hash(...))` as hash function at [here](https://github.com/apache/spark/pull/10703/files#diff-68c52246700eb099b0c896a3dacb5580R177)

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java [hunk 3]
```diff
@@ -290,9 +290,7 @@ public void testSetOperation() {
     List<String> data = Arrays.asList("abc", "abc", "xyz");
     Dataset<String> ds = context.createDataset(data, Encoders.STRING());
 
-    Assert.assertEquals(
-      Arrays.asList("abc", "xyz"),
-      sort(ds.distinct().collectAsList().toArray(new String[0])));
+    Assert.assertEquals(asSet("abc", "xyz"), toSet(ds.distinct().collectAsList()));
```
Comment (nongli) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java]: Same duplicate results thing. I think this should just sort instead of making into a set.

Comment (cloud-fan) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java]: `distinct` call will always produce distinct result rows right? I think it's safe to first convert to set then compare the results of aggregate/distinct.

Comment (nongli) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java]: Yea I guess this is fine. My point was that
x.toSet == y.toSet does not guarantee x.sorted == y.sorted. The verification is weaker.

**Comment by nongli** [comment]
LGTM

**Comment by cloud-fan** [comment]
retest this please.

**Comment by davies** [comment]
@cloud-fan Does this have some performance impact?

**Comment by cloud-fan** [comment]
@davies I think it will be faster, as now we have a codegen hash expression instead of `InternalRow.hashCode`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/bucket.scala [hunk 4]
```diff
@@ -56,3 +57,9 @@ private[sql] abstract class BucketedOutputWriterFactory extends OutputWriterFact
       context: TaskAttemptContext): OutputWriter =
     throw new UnsupportedOperationException("use bucket version")
 }
+
+private[sql] object BucketingUtils {
```
Comment (nongli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/bucket.scala]: it might be better to have this in HashPartition rather than BucketingUtils. HashPartitionionig seems like the more general concept.

**Comment by nongli** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/bucket.scala (hunk 4 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/bucket.scala]: +1 

Structurally it is better for bucketing to be a user of a general hash expression, rather than the other way around.

**Comment by cloud-fan** [comment]
@shivaram could you help to take a look at the failed R test? I tried many times but couldn't make it right...

**Comment by rxin** [comment]
@cloud-fan can you write a better pull request description?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala [hunk 5]
```diff
@@ -324,7 +324,7 @@ private[sql] class DynamicPartitionWriterContainer(
 
   private def bucketIdExpression: Option[Expression] = for {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: i think we prefer just using a simple map here

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 6]
```diff
@@ -1173,7 +1173,7 @@ test_that("group by, agg functions", {
 
   expect_equal(3, count(mean(gd)))
   expect_equal(3, count(max(gd)))
-  expect_equal(30, collect(max(gd))[1, 2])
+  expect_equal(30, collect(max(gd))[2, 3])
```
Comment (shivaram) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: I am not sure I understand why this line changed in this PR ? To debug this you can print the output with something like `debug <- collect(max(gd)); print(debug)`

Comment (cloud-fan) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: because the hash function in shuffle is changed in this PR, so the result of aggregate may have different order now.

Comment (cloud-fan) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: I'm not sure if I understand this test correctly, looks like it's trying to test the second row of the result of `max(gd)` to see if it's 30.

Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: it's picking the 1st row, 2nd column. eg.
```
> head(iris)
  Sepal.Length Sepal.Width Petal.Length Petal.Width Species
1          5.1         3.5          1.4         0.2  setosa
2          4.9         3.0          1.4         0.2  setosa
3          4.7         3.2          1.3         0.2  setosa
4          4.6         3.1          1.5         0.2  setosa
5          5.0         3.6          1.4         0.2  setosa
6          5.4         3.9          1.7         0.4  setosa
> iris[1,1]
[1] 5.1
> iris[1,2]
[1] 3.5
> iris[2,3]
[1] 1.4
```
I'm not sure if it should changed?

**Comment by rxin** [comment]
Going to merge this. Thanks.
